ci-operator/step-registry/user-setting/htpasswd/user-setting-htpasswd-commands.sh

```diff
@@ -0,0 +1,79 @@
+#!/bin/bash
+
+set -e
+set -u
+set -o pipefail
+
+function set_proxy () {
+    if test -s "${SHARED_DIR}/proxy-conf.sh" ; then
+        echo "setting the proxy"
+        # cat "${SHARED_DIR}/proxy-conf.sh"
+        echo "source ${SHARED_DIR}/proxy-conf.sh"
+        source "${SHARED_DIR}/proxy-conf.sh"
+    else
+        echo "no proxy setting."
+    fi
+}
+
+function set_users () {
+    # log in the cluster
+    if [ -f "${SHARED_DIR}/kubeconfig" ] ; then
+    export KUBECONFIG=${SHARED_DIR}/kubeconfig
+    fi
+
+    # prepare users
+    users=""
+    htpass_file=/tmp/users.htpasswd
+
+    for i in $(seq 1 50);
+    do
+        username="testuser-${i}"
+        password=$(< /dev/urandom tr -dc 'a-z0-9' | fold -w 12 | head -n 1 || true)
+        users+="${username}:${password},"
+        if [ -f "${htpass_file}" ]; then
+            htpasswd -B -b ${htpass_file} "${username}" "${password}"
+        else
+            htpasswd -c -B -b ${htpass_file} "${username}" "${password}"
+        fi
+    done
+
+    # current generation
+    gen=$(oc get deployment oauth-openshift -n openshift-authentication -o jsonpath='{.metadata.generation}')
+
+    # add users to cluster
+    oc create secret generic cucushift-htpass-secret --from-file=htpasswd=${htpass_file} -n openshift-config
+    oauth_file_src=/tmp/cucushift-oauth-src.yaml
+    oauth_file_dst=/tmp/cucushift-oauth-dst.yaml
+    oc get oauth cluster -o json > "${oauth_file_src}"
+    jq '.spec.identityProviders += [{"htpasswd":{"fileData":{"name":"cucushift-htpass-secret"}},"challenge":"true","login":"true","mappingMethod":"claim","name":"cucushift-htpasswd-provider","type":"HTPasswd"}]' "${oauth_file_src}" > "${oauth_file_dst}"
+    oc replace -f "${oauth_file_dst}"
+
+    # wait for oauth-openshift to rollout
+    wait_auth=true
+    expected_replicas=$(oc get deployment oauth-openshift -n openshift-authentication -o jsonpath='{.spec.replicas}')
+    while $wait_auth;
+    do
+        available_replicas=$(oc get deployment oauth-openshift -n openshift-authentication -o jsonpath='{.status.availableReplicas}')
+        new_gen=$(oc get deployment oauth-openshift -n openshift-authentication -o jsonpath='{.metadata.generation}')
+        if [[ $expected_replicas == "$available_replicas" && $((new_gen)) -gt $((gen)) ]]; then
+            wait_auth=false
+        else
+            sleep 10
+        fi
+    done
+
+    # store users in a shared file
+    if [ -f "${SHARED_DIR}/runtime_env" ] ; then
+    source "${SHARED_DIR}/runtime_env"
+    fi
+    runtime_env=${SHARED_DIR}/runtime_env
+    users=${users::-1}
+
+
+    cat <<EOF >>"${runtime_env}"
+export USERS=${users}
+EOF
```

Comment:
@jianzhangbjz , found above ci/rehearse/periodic-ci-openshift-verification-tests-master-nightly-4.11-upgrade-from-stable-4.10-from-stable-4.9-aws-ipi [logs](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_release/32626/rehearse-32626-periodic-ci-openshift-verification-tests-master-nightly-4.11-upgrade-from-stable-4.10-from-stable-4.9-aws-ipi/1574606738970120192/build-log.txt) show `USERS: unbound variable`:
```
'/tmp/kubeconfig-3084384126' -> '/tmp/kubeconfig'
/bin/bash: line 19: USERS: unbound variable
{"component":"entrypoint","error":"wrapped process failed: exit status 1","file":"k8s.io/test-infra/prow/entrypoint/run.go:80","func":"k8s.io/test-infra/prow/entrypoint.Options.Run","level":"error","msg":"Error executing test process","severity":"error","time":"2022-09-27T04:36:34Z"}
error: failed to execute wrapped command: exit status 1
```

Reply:
In this new step, seems you need explicitly run `export USERS=${users}` like originally `cucushift-pre` does.

Reply:
If you will have new update, could you indent line 67 BTW?

Reply:
Hi @xingxingxia , as you can see from https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_release/32626/rehearse-32626-periodic-ci-openshift-verification-tests-master-nightly-4.11-upgrade-from-stable-4.10-from-stable-4.9-aws-ipi/1574606738970120192/build-log.txt, this step `user-setting-htpasswd` not be run. So, we should update this `chain` to call this new step.

Reply:
The below error occurred when running https://github.com/openshift/release/blob/master/ci-operator/step-registry/cucushift/upgrade/prepare/cucushift-upgrade-prepare-commands.sh#L19
```yaml
[36mINFO[0m[2022-09-27T04:36:42Z] Logs for container test in pod aws-ipi-cucushift-upgrade-prepare: 
[36mINFO[0m[2022-09-27T04:36:42Z] '/tmp/kubeconfig-3084384126' -> '/tmp/kubeconfig'
/bin/bash: line 19: USERS: unbound variable
{"component":"entrypoint","error":"wrapped process failed: exit status 1","file":"k8s.io/test-infra/prow/entrypoint/run.go:80","func":"k8s.io/test-infra/prow/entrypoint.Options.Run","level":"error","msg":"Error executing test process","severity":"error","time":"2022-09-27T04:36:34Z"}
error: failed to execute wrapped command: exit status 1 
```

Reply:
The root cause is this `user-setting-htpasswd` weren't be called.

Reply:
That's weird! Not sure why this new step `user-setting-htpasswd` wasn't called. As you can see below, this `cucushift-upgrade-prepare` ran after the `cucushift-pre`, but it should be run after `user-setting-htpasswd`. I guess it's a flake. @danilo-gemoli Any suggestions? Thanks!
<img width="978" alt="image" src="https://user-images.githubusercontent.com/15416633/192924327-39fe8171-0749-4231-8795-af74c85a9802.png">

<img width="842" alt="image" src="https://user-images.githubusercontent.com/15416633/192924426-5e503180-3187-40a8-8263-95785168da8e.png">

Reply:
Another concern, is why not it failed at line 18?
<img width="442" alt="image" src="https://user-images.githubusercontent.com/15416633/192925786-da5ad30e-dca5-4dd1-9deb-83c363edc678.png">